### ISSUE TITLE
reference lateral join results at the top level

### ISSUE THREAD

**Issue by skeet70** [jawj/zapatos#111]
Disclaimer that I'm not very comfortable with lateral joins in general, so I may be missing something. Is there a way to select an extra column at the top level of a query that is a `COALESCE` of the results of lateral joins nested within that query? I'm not sure how else to explain the question, so I'll just leave an example of a query I currently have doing this:

```sql
select 
  "kms_configuration".*, 
  COALESCE(
    CAST(admin_user.id AS VARCHAR), 
    service_account.id
  ) AS created_by_id, 
from 
  "kms_configuration" 
  left join "admin_user" on "kms_configuration"."created_by" = "admin_user"."id" 
  left join "service_account" on "kms_configuration"."created_by_sa" = "service_account"."id" 
where 
  "kms_configuration"."id" = 415 
  and "kms_configuration"."deleted" = false 
  and "kms_configuration"."owner" = 1200 
limit 
  1
```

There will always be a `created_by` or `created_by_sa` on a configuration. I don't mind if the structure of the response is different, as long as the same data is there to be post processed, so I think I can accomplish this generally with lateral joins. The problem when I write that up is I can't figure out how to reproduce that `COALESCE` select. Is there a way to reference lateral joins in the top level `extras`?

```ts
db.selectExactlyOne("kms_configuration", {id: positiveInteger.unwrap(configId), deleted: false, owner: positiveInteger.unwrap(currentOrgId)}, {
            lateral: {
                admin_user: db.selectOne("admin_user", {id: db.parent("created_by")}),
                service_account: db.selectOne("service_account", { id: db.parent("created_by_sa") })
            },
            extras: {
                created_by_id: ???,
            }
        })
```

That's what I've got so far, but I'm unsure if I'm missing a feature that'd be useful.

**Comment by jawj** [comment]
I think the solution you've lighted on makes sense, but you're right that the numbered `ljoin` results make it brittle.

Looking at the code (shortcuts.ts#L513 and shortcuts.ts#L542), I can't immediately see any good reason why I shouldn't name these joins after the `lateral` name — i.e. use `ljoin_${raw(k)}` instead of `ljoin_${raw(String(i))}` on those two lines — which would make this quite a lot nicer.

Another solution to your immediate problem might be to wrap the complete `selectExactlyOne` query output in a brief custom raw `db.sql` query that adds the coalesced field you're looking for?

**Comment by jawj** [comment]
I just tried making [that change](https://github.com/jawj/zapatos/commit/e6e9b8fb74e80d9e087035b49fe220f56e94dc13) and nothing breaks in my demo or docs repos, so I may be tempted to roll that out before too long.

**Comment by skeet70** [comment]
Sorry, I saw this day of and forgot to respond!

Changing to have the name of the result in the lateral name would be great, that would immediately make it a lot more stable.

Wrapping with raw would be a way to solve this, but I'm trying to limit type holes to as local an area as possible.

**Comment by jawj** [comment]
OK, this change is now released in 5.0.1. Since the previous `ljoin_${n}` naming scheme was an undocumented implementation detail, I'm calling this a patch release.
